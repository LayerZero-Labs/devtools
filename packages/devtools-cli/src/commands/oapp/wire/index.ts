import { Command } from 'commander'
import { printLogo } from '@layerzerolabs/io-devtools/swag'
import { createLogger, setDefaultLogLevel } from '@layerzerolabs/io-devtools'
import {
    type WithLogLevelOption,
    type WithSetupOption,
    WithTsConfigOption,
    createLogLevelOption,
    createSetupFileOption,
    createTsConfigFileOption,
} from '@/commands/options'
import { setupTypescript } from '@/setup'

interface Args extends WithLogLevelOption, WithSetupOption, WithTsConfigOption {}

export const wire = new Command('wire')
    .addOption(createSetupFileOption())
    .addOption(createTsConfigFileOption())
    .addOption(createLogLevelOption())
    .action(async ({ setup, logLevel, tsConfig: tsConfigPath }: Args) => {
        printLogo()

        // We'll set the global logging level to get as much info as needed
        setDefaultLogLevel(logLevel)

        // We'll setup TypeScript support so that we can dynamically load TypeScript config files
        setupTypescript(tsConfigPath)

        const logger = createLogger(logLevel)

        logger.debug(`Loading setup from ${setup}`)

        logger.warn(
            `This command is just a placeholder. Please use @layerzerolabs/toolbox-hardhat package for the time being.`
        )
    })
