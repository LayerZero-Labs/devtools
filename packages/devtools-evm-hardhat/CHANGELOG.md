# @layerzerolabs/devtools-evm-hardhat

## 0.3.17

### Patch Changes

- 8b32a79: Use less strict version requirements for layerzero peer dependencies
- Updated dependencies [8b32a79]
  - @layerzerolabs/devtools-evm@0.3.10
  - @layerzerolabs/devtools@0.3.15
  - @layerzerolabs/export-deployments@0.0.9
  - @layerzerolabs/io-devtools@0.1.10

## 0.3.16

### Patch Changes

- aef7c85: Only load all deployments if a deployment could not be found

## 0.3.15

### Patch Changes

- 1e9b652: Exit with error code when simulation fails to start
- Updated dependencies [f3ab6c6]
  - @layerzerolabs/devtools@0.3.12

## 0.3.14

### Patch Changes

- 71d6101: Fix a display issue with deployment progressbar
- c190bf5: Add --stage CLI argument to the simulation task

## 0.3.13

### Patch Changes

- 938ac3d: Improve error reporting on UIntBigIntSchema
- Updated dependencies [938ac3d]
  - @layerzerolabs/io-devtools@0.1.9
  - @layerzerolabs/devtools@0.3.11
  - @layerzerolabs/devtools-evm@0.3.8
  - @layerzerolabs/export-deployments@0.0.8

## 0.3.12

### Patch Changes

- f865318: Update @LayerZero-Labs dependencies to 2.3.3
- Updated dependencies [f865318]
  - @layerzerolabs/export-deployments@0.0.7
  - @layerzerolabs/devtools-evm@0.3.7
  - @layerzerolabs/devtools@0.3.10

## 0.3.11

### Patch Changes

- 835cdbe: Replace Mumbai with Amoy

## 0.3.10

### Patch Changes

- c41c6be: Fix problems with missing deployments

## 0.3.9

### Patch Changes

- b20ba18: Add ability to use named signers
- Updated dependencies [b20ba18]
  - @layerzerolabs/devtools-evm@0.3.6

## 0.3.8

### Patch Changes

- 1047bcb: Add the ability to inherit task definitions

## 0.3.7

### Patch Changes

- cc080a9: Support local LayerZero environments

## 0.3.6

### Patch Changes

- 0f853ed: Adding `AsyncRetriable` to SDKs
- Updated dependencies [0f853ed]
- Updated dependencies [958ae04]
  - @layerzerolabs/devtools@0.3.6
  - @layerzerolabs/devtools-evm@0.3.5

## 0.3.5

### Patch Changes

- 7f2ebc6: Add --stage CLI argument to lz:deploy task

## 0.3.4

### Patch Changes

- 947a9aa: Add InferOmniGraph helper type
- 947a9aa: Add Configurator helper type
- Updated dependencies [947a9aa]
- Updated dependencies [947a9aa]
  - @layerzerolabs/devtools@0.3.5
  - @layerzerolabs/devtools-evm@0.3.4

## 0.3.3

### Patch Changes

- e16c864: Update @LayerZero-Labs dependencies to 2.1.27
- e16c864: Update executor config schema & types after update to 2.1.27
- Updated dependencies [e16c864]
- Updated dependencies [e16c864]
  - @layerzerolabs/export-deployments@0.0.6
  - @layerzerolabs/devtools-evm@0.3.3
  - @layerzerolabs/devtools@0.3.4

## 0.3.2

### Patch Changes

- c09680a: Update @LayerZero-Labs dependencies to 2.1.25
- Updated dependencies [c09680a]
  - @layerzerolabs/export-deployments@0.0.5
  - @layerzerolabs/devtools-evm@0.3.2
  - @layerzerolabs/devtools@0.3.3

## 0.3.1

### Patch Changes

- 059d817: Add the ability to choose gnosis safe for transaction signing; add the ability to pick signer index/address
- Updated dependencies [059d817]
  - @layerzerolabs/devtools-evm@0.3.1

## 0.3.0

### Minor Changes

- a4093ab: Adapt LayerZero package updates, including OApp version 2

### Patch Changes

- Updated dependencies [a4093ab]
  - @layerzerolabs/devtools-evm@0.3.0
  - @layerzerolabs/devtools@0.3.0

## 0.2.15

### Patch Changes

- 28b2598: Add simulation tasks under LZ_ENABLE_EXPERIMENTAL_SIMULATION feature flag

## 0.2.14

### Patch Changes

- eb4e4d0: Adjust simulation config logic for anvil
- ff6427c: Align the dependency ranges
- ff6427c: Update @LayerZero-Labs dependencies to 2.1.18
- Updated dependencies [f9987d3]
- Updated dependencies [ff6427c]
- Updated dependencies [ff6427c]
  - @layerzerolabs/devtools-evm@0.2.12
  - @layerzerolabs/devtools@0.2.10

## 0.2.13

### Patch Changes

- 622fd00: Add docker compose utilities for simulation

## 0.2.12

### Patch Changes

- 711cb98: Add config utilities for simulation
- 6d2e1f8: Add types for simulation
- Updated dependencies [6d2e1f8]
  - @layerzerolabs/devtools-evm@0.2.10
  - @layerzerolabs/devtools@0.2.9

## 0.2.11

### Patch Changes

- 54cf16e: Silence bigint-buffer warning
- Updated dependencies [54cf16e]
- Updated dependencies [2b9ae6a]
- Updated dependencies [2b9ae6a]
  - @layerzerolabs/devtools@0.2.7
  - @layerzerolabs/devtools-evm@0.2.8
  - @layerzerolabs/export-deployments@0.0.4
  - @layerzerolabs/io-devtools@0.1.5

## 0.2.10

### Patch Changes

- 9e78abe: Update hardhat-deploy
- Updated dependencies [9e78abe]
  - @layerzerolabs/io-devtools@0.1.4

## 0.2.9

### Patch Changes

- d0d1cb3: Mark hardhat tasks with exit code 1 if they were not successful

## 0.2.8

### Patch Changes

- e7ef1aa: Update @layerzero-labs dependencies to 2.1.15
- Updated dependencies [e7ef1aa]
  - @layerzerolabs/devtools-evm@0.2.7
  - @layerzerolabs/devtools@0.2.6

## 0.2.7

### Patch Changes

- b93a018: Update @layerzero-labs dependencies to 2.1.13
- Updated dependencies [b93a018]
  - @layerzerolabs/devtools-evm@0.2.6
  - @layerzerolabs/devtools@0.2.5

## 0.2.6

### Patch Changes

- 855fa36: Update to latest layerzerolabsmonorepo dependencies
- Updated dependencies [855fa36]
  - @layerzerolabs/export-deployments@0.0.3
  - @layerzerolabs/devtools-evm@0.2.5
  - @layerzerolabs/devtools@0.2.4

## 0.2.5

### Patch Changes

- 2732dfc: Add support for submitting transactions to Gnosis Safes using @safe-global
- Updated dependencies [2732dfc]
  - @layerzerolabs/devtools-evm@0.2.4

## 0.2.4

### Patch Changes

- f0392c5: Add export deployments hardhat task

## 0.2.3

### Patch Changes

- e0f41b5: Update @layerzero-labs dependencies
- Updated dependencies [e0f41b5]
  - @layerzerolabs/devtools-evm@0.2.3
  - @layerzerolabs/devtools@0.2.3

## 0.2.2

### Patch Changes

- 491b5a5: Upgrade contract Dependencies to 2.1.7
- Updated dependencies [491b5a5]
  - @layerzerolabs/devtools-evm@0.2.2
  - @layerzerolabs/devtools@0.2.2

## 0.2.1

### Patch Changes

- 71e355c: Update "@layerzerolabs/\*" dependencies
- Updated dependencies [71e355c]
  - @layerzerolabs/devtools-evm@0.2.1
  - @layerzerolabs/devtools@0.2.1

## 0.2.0

### Minor Changes

- 6e464f2: Specify EndpointV2 with the "V2" suffix wherever appropriate

### Patch Changes

- Updated dependencies [6e464f2]
  - @layerzerolabs/devtools-evm@0.2.0
  - @layerzerolabs/devtools@0.2.0

## 0.1.6

### Patch Changes

- 700168d: Move sign & send logic into a subtask

## 0.1.5

### Patch Changes

- 5675416: Fix problems with missing deployments when running lz:deploy

## 0.1.4

### Patch Changes

- 8227742: Update @layerzero-labs dependencies to 2.1.4
- 3a438b1: Compile project before depolying in lz:deploy
- Updated dependencies [8227742]
  - @layerzerolabs/devtools-evm@0.1.3
  - @layerzerolabs/devtools@0.1.4

## 0.1.3

### Patch Changes

- 5c58d69: Update @layerzerolabs dependencies to 2.1.3
- 253c79e: Add lz:deploy task; Add createClearDeployments utility
- Updated dependencies [253c79e]
- Updated dependencies [5c58d69]
  - @layerzerolabs/io-devtools@0.1.2
  - @layerzerolabs/devtools-evm@0.1.2
  - @layerzerolabs/devtools@0.1.3

## 0.1.2

### Patch Changes

- 40e8dba: Remove CommaSeparatedValuesSchema in favor of splitCommaSeparated; Move LogLevelSchema; Export isLogLevel
- Updated dependencies [40e8dba]
- Updated dependencies [4258ef3]
  - @layerzerolabs/io-devtools@0.1.1
  - @layerzerolabs/devtools@0.1.2

## 0.1.1

### Patch Changes

- f0036c5: Use monorepo 2.1.2 released dependencies
- Updated dependencies [f0036c5]
  - @layerzerolabs/devtools@0.1.1
  - @layerzerolabs/devtools-evm@0.1.1

## 0.1.0

### Minor Changes

- 120adf1: Make packages public

### Patch Changes

- Updated dependencies [120adf1]
  - @layerzerolabs/devtools-evm@0.1.0
  - @layerzerolabs/io-devtools@0.1.0
  - @layerzerolabs/devtools@0.1.0

## 0.0.11

### Patch Changes

- 17c8a23: Fix problems with --networks hardhat CLI argument parser

## 0.0.10

### Patch Changes

- 0877186: Update @layerzerolabs dependencies to 2.0.25 and 2.0.26-rc1
- Updated dependencies [0877186]
  - @layerzerolabs/devtools-evm@0.0.6
  - @layerzerolabs/devtools@0.0.5

## 0.0.9

### Patch Changes

- e5a9f35: Add hardhat CLI argument parsers for networks and for general CSV values

## 0.0.8

### Patch Changes

- 565b646: Allow project packages to specify the external artifacts and deployments in their hardhat config
- Updated dependencies [2b36ed9]
- Updated dependencies [2b36ed9]
  - @layerzerolabs/io-devtools@0.0.6

## 0.0.7

### Patch Changes

- 4318721: Fix an issue where transient artifacts would not be included in error parsing
- 4318721: Add signature to the output of lz:errors:list task
- Updated dependencies [8931fb2]
  - @layerzerolabs/devtools-evm@0.0.5

## 0.0.6

### Patch Changes

- 56706ff: Add a feature-flagged lz:deploy hasrhat task

## 0.0.5

### Patch Changes

- 1e375c9: Adding EVM testing abilities for hardhat and foundry

## 0.0.4

### Patch Changes

- af8cc25: Update dependencies
- Updated dependencies [af8cc25]
  - @layerzerolabs/devtools-evm@0.0.4
  - @layerzerolabs/io-devtools@0.0.4
  - @layerzerolabs/devtools@0.0.4

## 0.0.3

### Patch Changes

- 2183cfc: Add lz:errors:list task
- 2183cfc: Add getAllArtifacts helper
- fe1542e: Add getEidsByNetworkName utility
- Updated dependencies [5077bf1]
- Updated dependencies [5077bf1]
  - @layerzerolabs/io-devtools@0.0.3
  - @layerzerolabs/devtools-evm@0.0.3

## 0.0.2

### Patch Changes

- 5236166: Add a generic contract error parser for hardhat projects
- 6483544: Update dependency versions
- 70646b4: Add missing dependencies
- 8ac903c: Fix type extension imports for toolbox-hardhat
- Updated dependencies [e0be8b7]
- Updated dependencies [70646b4]
- Updated dependencies [b5991ca]
- Updated dependencies [2dac0da]
  - @layerzerolabs/io-devtools@0.0.2
  - @layerzerolabs/devtools-evm@0.0.2
  - @layerzerolabs/devtools@0.0.2

## 0.0.1

### Patch Changes

- b74afbe: Initial 0.0.1 version
- Updated dependencies [b74afbe]
  - @layerzerolabs/devtools-evm@0.0.1
  - @layerzerolabs/io-devtools@0.0.1

## 0.0.2

### Patch Changes

- 6964deb: Memoize create\* calls
