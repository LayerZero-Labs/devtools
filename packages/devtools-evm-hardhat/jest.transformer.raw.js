// This is a quick, up-to-date version of https://www.npmjs.com/package/jest-raw-loader
//
// We need this in order to be able to statically import/embed simulation Dockerfile & nginx.conf
module.exports = {
    process: (content) => ({ code: `module.exports = ${JSON.stringify(content)}` }),
};
