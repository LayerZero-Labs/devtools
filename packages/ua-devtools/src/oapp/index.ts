export * from './check'
export * from './config'
export * from './schema'
export * from './types'
