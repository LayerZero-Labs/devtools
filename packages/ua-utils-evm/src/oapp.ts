import { Contract } from "@ethersproject/contracts"
import { GetPropertyValue, createProperty } from "@layerzerolabs/ua-utils"
import { EndpointId } from "@layerzerolabs/lz-definitions"

type SetPeerConfigurableContext = [oapp: Contract, endpointId: EndpointId]

type SetPeerConfigurableValue = string

export const createSetPeerProperty = (desired: GetPropertyValue<SetPeerConfigurableContext, SetPeerConfigurableValue>) =>
    createProperty<SetPeerConfigurableContext, SetPeerConfigurableValue>({
        desired,
        get: (oapp, endpointId) => oapp.peers(endpointId),
        set: (oapp, endpointId, peer) => oapp.setPeer(endpointId, peer),
    })
