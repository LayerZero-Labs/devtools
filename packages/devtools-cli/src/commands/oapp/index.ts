import { Command } from 'commander'
import { wire } from './wire'

export const oapp = new Command('oapp').description('OApp configuration commands').addCommand(wire)
