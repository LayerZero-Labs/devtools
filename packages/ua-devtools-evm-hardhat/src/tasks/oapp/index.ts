import './wire'
import './peers.get'
import './config.get.default'
import './config.get.executor'
import './config.get'
import './enforced.opts.get'
import './config.init'
import './subtask.config.load'
