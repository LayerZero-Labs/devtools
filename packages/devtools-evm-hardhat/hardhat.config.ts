import { EndpointId } from '@layerzerolabs/lz-definitions'
import 'hardhat-deploy'
import { HardhatUserConfig } from 'hardhat/types'

// We import the locally defined tasks
import './src/tasks'

/**
 * This is a dummy hardhat config that enables us to test
 * hardhat functionality without mocking too much
 */
const config: HardhatUserConfig = {
    networks: {
        'ethereum-mainnet': {
            url: 'no:///way',
            saveDeployments: false,
            eid: EndpointId.ETHEREUM_V2_MAINNET,
        },
        'ethereum-testnet': {
            url: 'no:///way',
            saveDeployments: false,
            eid: EndpointId.ETHEREUM_TESTNET,
        },
        'bsc-testnet': {
            url: 'no:///way',
            saveDeployments: false,
            accounts: {
                mnemonic: 'test test test test test test test test test test test junk',
            },
        },
    },
}

export default config
