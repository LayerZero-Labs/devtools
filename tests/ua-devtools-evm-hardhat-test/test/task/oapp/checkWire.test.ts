import hre from 'hardhat'
import { resolve } from 'path'
import { isFile } from '@layerzerolabs/io-devtools'
import { deployOAppFixture } from '../../__utils__/oapp'
import { EndpointId } from '@layerzerolabs/lz-definitions'
import { createContractFactory, createSignerFactory } from '@layerzerolabs/devtools-evm-hardhat'
import { omniContractToPoint } from '@layerzerolabs/devtools-evm'
import { createOAppFactory } from '@layerzerolabs/ua-devtools-evm'
import { TASK_LZ_CHECK_WIRE_OAPP } from '@layerzerolabs/ua-devtools-evm-hardhat'

describe.only('task: checkWire', () => {
    const CONFIGS_BASE_DIR = resolve(__dirname, '__data__', 'configs')
    const configPathFixture = (fileName: string): string => {
        const path = resolve(CONFIGS_BASE_DIR, fileName)
        expect(isFile(path)).toBeTruthy()
        return path
    }

    beforeEach(async () => {
        await deployOAppFixture()
    })

    it('should show no chains are connected', async () => {
        const oappConfig = configPathFixture('valid.config.connected.js')
        await expect(hre.run(TASK_LZ_CHECK_WIRE_OAPP, { oappConfig })).resolves.toMatchSnapshot()
    })

    it('should show one chain (eth) is connected', async () => {
        const oappConfig = configPathFixture('valid.config.connected.js')
        const ethContract = {
            eid: EndpointId.ETHEREUM_MAINNET,
            contractName: 'DefaultOApp',
        }
        const avaxContract = {
            eid: EndpointId.AVALANCHE_MAINNET,
            contractName: 'DefaultOApp',
        }
        const contractFactory = createContractFactory()
        const sdkFactory = createOAppFactory(contractFactory)
        const ethPoint = omniContractToPoint(await contractFactory(ethContract))
        const ethOAppSdk = await sdkFactory(ethPoint)
        const avaxPoint = omniContractToPoint(await contractFactory(avaxContract))
        {
            const signerFactory = createSignerFactory()
            const ethSigner = await signerFactory(ethContract.eid)
            await ethSigner.signAndSend(await ethOAppSdk.setPeer(avaxPoint.eid, avaxPoint.address))
        }
        await expect(hre.run(TASK_LZ_CHECK_WIRE_OAPP, { oappConfig })).resolves.toMatchSnapshot()
    })

    it('should show all chains are connected', async () => {
        const oappConfig = configPathFixture('valid.config.connected.js')
        const ethContract = {
            eid: EndpointId.ETHEREUM_MAINNET,
            contractName: 'DefaultOApp',
        }
        const avaxContract = {
            eid: EndpointId.AVALANCHE_MAINNET,
            contractName: 'DefaultOApp',
        }
        const contractFactory = createContractFactory()
        const sdkFactory = createOAppFactory(contractFactory)

        const ethPoint = omniContractToPoint(await contractFactory(ethContract))
        const ethOAppSdk = await sdkFactory(ethPoint)
        const avaxPoint = omniContractToPoint(await contractFactory(avaxContract))
        const avaxOAppSdk = await sdkFactory(avaxPoint)
        {
            const signerFactory = createSignerFactory()

            const ethSigner = await signerFactory(ethContract.eid)
            await ethSigner.signAndSend(await ethOAppSdk.setPeer(avaxPoint.eid, avaxPoint.address))

            const avaxSigner = await signerFactory(avaxContract.eid)
            await avaxSigner.signAndSend(await avaxOAppSdk.setPeer(ethPoint.eid, ethPoint.address))
        }
        await expect(hre.run(TASK_LZ_CHECK_WIRE_OAPP, { oappConfig })).resolves.toMatchSnapshot()
    })
})
