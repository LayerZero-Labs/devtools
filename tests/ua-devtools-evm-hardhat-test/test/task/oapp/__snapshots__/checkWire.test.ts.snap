// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`task: checkWire should show all chains are connected 1`] = `
[
  {
    "hasPeer": true,
    "vector": {
      "from": {
        "address": "0x38a024C0b412B9d1db8BC398140D00F5Af3093D4",
        "eid": 30106,
      },
      "to": {
        "address": "0x38a024C0b412B9d1db8BC398140D00F5Af3093D4",
        "eid": 30101,
      },
    },
  },
  {
    "hasPeer": true,
    "vector": {
      "from": {
        "address": "0x38a024C0b412B9d1db8BC398140D00F5Af3093D4",
        "eid": 30101,
      },
      "to": {
        "address": "0x38a024C0b412B9d1db8BC398140D00F5Af3093D4",
        "eid": 30106,
      },
    },
  },
]
`;

exports[`task: checkWire should show no chains are connected 1`] = `
[
  {
    "hasPeer": false,
    "vector": {
      "from": {
        "address": "0x38a024C0b412B9d1db8BC398140D00F5Af3093D4",
        "eid": 30106,
      },
      "to": {
        "address": "0x38a024C0b412B9d1db8BC398140D00F5Af3093D4",
        "eid": 30101,
      },
    },
  },
  {
    "hasPeer": false,
    "vector": {
      "from": {
        "address": "0x38a024C0b412B9d1db8BC398140D00F5Af3093D4",
        "eid": 30101,
      },
      "to": {
        "address": "0x38a024C0b412B9d1db8BC398140D00F5Af3093D4",
        "eid": 30106,
      },
    },
  },
]
`;

exports[`task: checkWire should show one chain (eth) is connected 1`] = `
[
  {
    "hasPeer": false,
    "vector": {
      "from": {
        "address": "0x38a024C0b412B9d1db8BC398140D00F5Af3093D4",
        "eid": 30106,
      },
      "to": {
        "address": "0x38a024C0b412B9d1db8BC398140D00F5Af3093D4",
        "eid": 30101,
      },
    },
  },
  {
    "hasPeer": true,
    "vector": {
      "from": {
        "address": "0x38a024C0b412B9d1db8BC398140D00F5Af3093D4",
        "eid": 30101,
      },
      "to": {
        "address": "0x38a024C0b412B9d1db8BC398140D00F5Af3093D4",
        "eid": 30106,
      },
    },
  },
]
`;
